File: app/src/main/java/com/travlog/android/apps/ui/activities/SetUsernameActivity.java
package com.travlog.android.apps.ui.activities;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.jakewharton.rxbinding2.view.RxView;
import com.jakewharton.rxbinding2.widget.RxTextView;
import com.travlog.android.apps.R;
import com.travlog.android.apps.libs.BaseActivity;
import com.travlog.android.apps.libs.qualifiers.RequiresActivityViewModel;
import com.travlog.android.apps.viewmodels.SetUsernameViewModel;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;

@RequiresActivityViewModel(SetUsernameViewModel.class)
public class SetUsernameActivity extends BaseActivity<SetUsernameViewModel> {

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.username_input)
    TextInputLayout usernameInputLayout;
    @BindView(R.id.username_edit)
    TextInputEditText usernameEdit;
    @BindView(R.id.save_button)
    View saveButton;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.a_set_username);
        setSupportActionBar(toolbar);
        setDisplayHomeAsUpEnabled(true);

        RxTextView.textChanges(usernameEdit)
                .map(CharSequence::toString)
                .compose(bindToLifecycle())
                .subscribe(viewModel.inputs::username);

        RxView.clicks(saveButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> viewModel.saveClick());

        viewModel.outputs.setSaveButtonEnabled()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setSaveButtonEnabled);

        viewModel.outputs.back()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::back);
    }

    private void setSaveButtonEnabled(final boolean enabled) {
        this.saveButton.setEnabled(enabled);
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/activities/SetUsernameActivity.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.ui.activities

import android.os.Bundle
import android.util.Pair
import com.jakewharton.rxbinding3.view.clicks
import com.jakewharton.rxbinding3.widget.textChanges
import com.travlog.android.apps.R
import com.travlog.android.apps.ViewModelFactory
import com.travlog.android.apps.getAppInjector
import com.travlog.android.apps.getViewModel
import com.travlog.android.apps.libs.BaseActivity
import com.travlog.android.apps.libs.utils.TransitionUtils.slideInFromLeft
import com.travlog.android.apps.viewmodels.SetUsernameViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.a_set_username.*
import javax.inject.Inject

class SetUsernameActivity : BaseActivity<SetUsernameViewModel>() {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    override fun onCreate(savedInstanceState: Bundle?) {
        getAppInjector().inject(this)
        viewModel = getViewModel(viewModelFactory)

        super.onCreate(savedInstanceState)

        setContentView(R.layout.a_set_username)
        setSupportActionBar(toolbar)
        setDisplayHomeAsUpEnabled(true)

        username_edit.textChanges()
                .map { it.toString() }
                .compose(bindToLifecycle())
                .subscribe(viewModel!!.inputs::username)

        done.clicks()
                .compose(bindToLifecycle())
                .subscribe { viewModel!!.doneClick() }

        viewModel!!.outputs.setDoneButtonEnabled()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setDoneButtonEnabled)

        addDisposable(
                viewModel!!.outputs.back()
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe { this.back() })
    }

    private fun setDoneButtonEnabled(enabled: Boolean) {
        this.done.isEnabled = enabled
    }

    override fun exitTransition(): Pair<Int, Int>? {
        return slideInFromLeft()
    }
}
